### PR TITLE
load custom palette in editor automatically

### LINKED ISSUE #315: Support custom palette in editor and resource files
**Is your feature request related to a problem? Please describe.**
We can run a pyxel game with a custom palette, but we can’t use it in the editor.

**Describe the solution you'd like**
Idea 1: add a command-line parameter to the editor like `--palette 1a1c2c,5d275d,b13e53,ef7d57`
This could be a little unconvenient to use though.

Idea 2: add a param like `--palette path/to/file` that reads the palette from a file.  Could use a text file format (lospec for example has multiple text formats for each palette) for easy parsing, or a PNG file with 1px per colour but that would require PIL or pypng to analyse.  I have a clone of the repo and added this option for my own usage, so I can send a PR.

Idea 3: update resource file format to store palette in the resource file.  Load automatically when opening a resource file with a custom palette.  To save a custom palette, either keep it a secret feature (people have to add palette file to the zip) or add a palette tool in editor.  I probably can’t make a PR for this, and you’re doing the rust rewrite so probably better to wait for that.

**Describe alternatives you've considered**
See above

**Additional context**

### PR DESCRIPTION
**Pull request by merwok**
Fixes #315
Rework of #321

todo:
- document
- add constant for pyxpal extension

### CODE REVIEW

File: pyxel/editor/app.py [hunk 1]
```diff
@@ -27,14 +27,23 @@ class App(Widget):
...
         pyxel.mouse(True)
         self._set_title(resource_file)
+
         if os.path.exists(resource_file):
             pyxel.load(resource_file)
 
+        if os.path.exists(palette_file):
+            with open(palette_file) as file:
+                palette = [int(line.lstrip("#"), 16)
+                           for line in file.read().splitlines()]
```
Comment (merwok) [inline pyxel/editor/app.py:44]: I could improve error reporting in case of bad file format.

Comment (kitao) [reply pyxel/editor/app.py:44]: Pyxel is also designed for working as Rust module.
So palette-related things should be implemented for pyxel.load as Rust code not Python.

Comment (merwok) [reply pyxel/editor/app.py:44]: But the asset loading is there in the Python code!  The editor is a Python app, not a Rust one, correct?

**Comment by kitao** [comment]
No, asset loading is the role of load function. Editor just use it. So if
user call load function, they can expect the same loading mechanism
including palette management.

2022年1月12日(水) 12:08 Éric ***@***.***>:

**Comment by merwok** [comment]
Ah, I understand.

Do you approve the basic ideas of this PR?
- filename is same as assets -pyxres +pyxpal
- format is text file, 16 lines with `rrggbb` or `#rrggbb`

Or would it be better to have the palette as a new kind of data inside the assets file, make it editable in the editor somehow, and automatically applied in game?

**Comment by kitao** [comment]
The specs you mentioned are almost the same as my idea.
If the palette is included in the .pyxres, it means Pyxel's editor should support palette editing tool, but It takes too much time only for the small number of users to use the function.

There are still a few things to clarify the design.
- In the first place, for Pyxel, users who use custom palettes are positioned as users who have the skills to implement the functions themselves, so how much can Pyxel complicate the functions for them?
- If there's a user who just would like to use .pyxpal without .pyxres. What kind of solution should be provided?
- If other external settings such as touch panel handling way are required in the future, how to handle them. It can be better that Pyxel support some setting file and palette information is one of the entries

From time to time I get a request to increase the number of colors, but I reject it because it goes against Pyxel's idea of keeping it simple and retro. I think I first need to make everyone understand that custom palettes deviate by half that policy.

**Comment by merwok** [comment]
>  It takes too much time only for the small number of users to use the function.

OK, so good reason to **not** add palette inside pyxres file.

> In the first place, for Pyxel, users who use custom palettes are positioned as users who have the skills to implement the functions themselves, so how much can Pyxel complicate the functions for them?

Could you rephrase?  I don’t understand this (you want pyxel to make some functions more complicated?)

>  If there's a user who just would like to use .pyxpal without .pyxres. What kind of solution should be provided?

There’s no use case.  If you are not using pyxpal with pyxres, then you are not using the assets editor.

If you have a custom palette for your game and you load images directly, then this PR does not apply, you can already use `pyxel.colors.from_list` in your game code.

> If other external settings such as touch panel handling way are required in the future, how to handle them.
> It can be better that Pyxel support some setting file and palette information is one of the entries

I do not know.

> […] Pyxel's idea of keeping it simple and retro. I think I first need to make everyone understand that custom palettes deviate by half that policy.

16 colours is great!  But I would say again that we do have support for custom palette in game code, this is just making the assets editor more convenient.

**Comment by kitao** [comment]
Sorry for my imperfect English skill.

What I would like to say in the first bullet is that I think there are two types of users.

One is programming learner and students, and I'm careful not to over-complicate the functionality, not to increase the amount of manuals, to make it as easy to use as possible for them.

The other is programming-skilled users. For them, I give Pyxel some flexibility, but try not to cover the features they can create themselves. Because Pyxel is not a professional tool and there are many other game engines for that purpose. That is the Pyxel's basic design policy.

And I think that most of the people who use custom palettes are advanced programmers, so simply adding the necessary functions for that is not suitable for the policy. Therefore, I think it is necessary to think about what functions and flexibility are really necessary for the custom palette, and to think about how to cover it to the minimum, instead of adding it because Pyxel doesn't have that feature.

**Comment by merwok** [comment]
I understand your point about keeping Pyxel accessible for novice programmers.

But I feel that you don’t get my point: we can already use an advanced function to change the palette in game, but we can’t change the palette in the assets editor.

**Comment by kitao** [comment]
No, what would like to say is we should think about what kind of way is
suitable for Pyxel carefully.

2022年1月13日(木) 0:20 Éric ***@***.***>:

**Comment by merwok** [comment]
Ah, I understand and fully agree!

I can wait until you have a design that you like.  The palette could be inside pyxres, or separate file that’s automatically loaded, or manually loaded, or part of a settings file…

**Comment by kitao** [comment]
Thank you for your understanding. But you don’t need to wait for my
thoughts to settle. Any suggestion is very welcome if it is along with the
policy!

2022年1月13日(木) 0:30 Éric ***@***.***>:

**Comment by merwok** [comment]
This PR is one proposal: load palette in editor.
You countered: load palette in rust code when loading asset file.
Should I (try to) implement your idea?

**Comment by kitao** [comment]
No, you don't need to implement it (yet at let).
Now I'm thinking about what happens if Pyxel has common setting file and it is loaded when Pyxel is initialized and loading resource.

**Comment by merwok** [comment]
Any update on your thinking for this?  🙂

**Comment by matteoturini** [comment]
> And I think that most of the people who use custom palettes are advanced programmers
and colorblind people as I have mentioned in the issue
